### PR TITLE
Attention processor cross attention norm group norm

### PR DESCRIPTION
**Pull request by williamberman**
This lets the cross attention norm use both a group norm block or a
layer norm block.

The group norm operates along the channels dimension
and requires input shape (batch size, channels, *) where as the layer norm with a single
`normalized_shape` dimension only operates over the least significant
dimension i.e. (*, channels).

The channels we want to normalize along are the hidden dimension of the encoder hidden states.

By convention, the encoder hidden states are always passed as (batch size, sequence
length, hidden states).

This means the layer norm can operate on the tensor without modification, but the group
norm requires flipping the last two dimensions to operate on (batch size, hidden states, sequence length).

All existing attention processors will have the same logic and we can
consolidate it in a helper function `prepare_encoder_hidden_states`

this is rebased on top of https://github.com/huggingface/diffusers/pull/3014 because it requires the bug fix of making sure the added kv processors always take encoder states as (batch size, sequence
length, hidden states)

### CODE REVIEW

File: src/diffusers/models/attention_processor.py [hunk 1]
```diff
@@ -68,7 +69,6 @@ def __init__(
         cross_attention_dim = cross_attention_dim if cross_attention_dim is not None else query_dim
         self.upcast_attention = upcast_attention
         self.upcast_softmax = upcast_softmax
-        self.cross_attention_norm = cross_attention_norm
```
Comment (williamberman) [inline src/diffusers/models/attention_processor.py:71]: If a layer's existence is configured with a boolean flag, we usually define the layer as none or defined, and then check if the layer is not none in the forward method. I consolidated this check in the `prepare_encoder_hidden_states` method and removed all references to the boolean

File: src/diffusers/models/attention_processor.py [hunk 2]
```diff
@@ -56,7 +56,8 @@ def __init__(
         bias=False,
         upcast_attention: bool = False,
         upcast_softmax: bool = False,
-        cross_attention_norm: bool = False,
+        cross_attention_norm: Union[bool, str] = False,
```
Comment (williamberman) [inline src/diffusers/models/attention_processor.py:59]: I'd prefer to not do both a boolean or string, but this is the simplest method to upgrade the flag to support multiple layer types

File: src/diffusers/models/attention_processor.py [hunk 3]
```diff
@@ -56,7 +56,8 @@ def __init__(
...
+        cross_attention_norm_num_groups: int = 32,
```
Comment (williamberman) [inline src/diffusers/models/attention_processor.py:60]: iirc, group norms in attention blocks for ldm codebases/variants almost always use 32 groups. We probably won't have to actually set this value often

File: src/diffusers/models/attention_processor.py [hunk 4]
```diff
@@ -291,6 +311,29 @@ def prepare_attention_mask(self, attention_mask, target_length, batch_size=None)
...
+            # (batch_size, sequence_length, hidden_size) ->
+            # (batch_size, hidden_size, sequence_length)
+            encoder_hidden_states = encoder_hidden_states.transpose(1, 2)
+            encoder_hidden_states = self.norm_cross(encoder_hidden_states)
+            encoder_hidden_states = encoder_hidden_states.transpose(1, 2)
+        else:
+            assert False
+
+        return encoder_hidden_states
+
```
Comment (williamberman) [inline src/diffusers/models/attention_processor.py:336]: Standardized encoder hidden states preprocessing

File: src/diffusers/models/attention_processor.py [hunk 5]
```diff
@@ -56,7 +56,8 @@ def __init__(
         bias=False,
         upcast_attention: bool = False,
         upcast_softmax: bool = False,
-        cross_attention_norm: bool = False,
+        cross_attention_norm: Optional[Union[bool, str]] = False,
```
Comment (williamberman) [inline src/diffusers/models/attention_processor.py:59]: Ideally, this should just be an `Optional[str]` type now but this is backwards compatible. In the unet and unet blocks, I treat the type as `Optional[str]` and it's still pretty clean

File: src/diffusers/models/attention_processor.py [hunk 6]
```diff
@@ -56,7 +56,8 @@ def __init__(
         bias=False,
         upcast_attention: bool = False,
         upcast_softmax: bool = False,
-        cross_attention_norm: bool = False,
+        cross_attention_norm: Optional[str] = None,
```
Comment (williamberman) [inline src/diffusers/models/attention_processor.py:59]: This can be safely upgraded to an `Optional[str]` as we only ever use it internally through the "K attention" blocks which are updated in this PR

File: src/diffusers/models/attention_processor.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/diffusers/models/attention_processor.py:71]: ok for me

File: src/diffusers/models/attention_processor.py [hunk 7]
```diff
@@ -291,6 +311,29 @@ def prepare_attention_mask(self, attention_mask, target_length, batch_size=None)
             attention_mask = attention_mask.repeat_interleave(head_size, dim=0)
         return attention_mask
 
+    def prepare_encoder_hidden_states(self, hidden_states, encoder_hidden_states=None):
```
Comment (patrickvonplaten) [inline src/diffusers/models/attention_processor.py:314]: not a huge fan of the naming here. Can we call it maybe instead:

```suggestion
    def norm_encoder_hidden_states(self, hidden_states, encoder_hidden_states=None):
```

**Review by patrickvonplaten — APPROVED** [review]
Ok for me, but let's maybe give it a different name.
Also, let's maybe make sure to not factor out to much logic as it makes the function much harder to read

File: src/diffusers/models/attention_processor.py (hunk 7 above)
Comment (williamberman) [reply src/diffusers/models/attention_processor.py]: renamed here and moved to the check on if encoder_hidden_states is None back to the attention processors to better reflect the new name https://github.com/huggingface/diffusers/pull/3021/commits/9160e517b3bf8ee0b6ca4923c49c928f2a657ef9

File: src/diffusers/models/attention_processor.py (hunk 4 above)
Comment (sayakpaul) [reply src/diffusers/models/attention_processor.py]: Was this method renamed to `norm_encoder_hidden_states()`?

Comment (sayakpaul) [reply src/diffusers/models/attention_processor.py]: Nevermind. 

https://github.com/huggingface/diffusers/pull/3021/commits/9160e517b3bf8ee0b6ca4923c49c928f2a657ef9

**Review by sayakpaul — APPROVED** [review]

**Review by yiyixuxu — APPROVED** [review]
nice :)

File: src/diffusers/models/attention_processor.py (hunk 4 above)
Comment (williamberman) [reply src/diffusers/models/attention_processor.py]: yessir!
